thorlcr/master/thmastermain.cpp

```diff
@@ -441,9 +441,61 @@ bool checkClusterRelicateDAFS(IGroup &grp)
 
 static bool auditStartLogged = false;
 
+class CThorEndHandler : public CSimpleInterface, implements IThreaded
+{
+    CThreaded threaded;
+    unsigned timeout;
+    std::atomic<bool> started{false};
+    std::atomic<bool> stopped{false};
+    Semaphore sem;
+public:
+    CThorEndHandler() : threaded("CThorEndHandler"), timeout(30000)
+    {
+        threaded.init(this);
+    }
+    ~CThorEndHandler()
+    {
+        stop();
+        threaded.join(timeout);
+    }
+    void start(unsigned timeoutSecs)
+    {
+        bool expected = false;
+        if (started.compare_exchange_strong(expected, true));
+        {
+            timeout = timeoutSecs * 1000; // sem_post and sem_wait are mem_barriers
+            sem.signal();
+        }
+    }
+    void stop()
+    {
+        bool expected = false;
+        if (stopped.compare_exchange_strong(expected, true));
+        {
+            sem.signal();
+        }
```

Comment:
style: normally we don't put 1 line blocks in braces.

Reply:
done